demos/defender/aws_iot_demo_defender.c

```diff
@@ -154,138 +92,216 @@ void _defenderCallback( void * param1,
+    IotMqttConnection_t mqttConnection = IOT_MQTT_CONNECTION_INITIALIZER;
 
-    /* Set network information. */
-    startInfo.mqttNetworkInfo = ( IotMqttNetworkInfo_t ) IOT_MQTT_NETWORK_INFO_INITIALIZER;
-    startInfo.mqttNetworkInfo.createNetworkConnection = true;
-    startInfo.mqttNetworkInfo.u.setup.pNetworkServerInfo = &_DEFENDER_SERVER_INFO;
-    startInfo.mqttNetworkInfo.u.setup.pNetworkCredentialInfo = &_AWS_IOT_CREDENTIALS;
+    /* Unused parameters. */
+    ( void ) awsIotMqttMode;
 
-    /* Only set ALPN protocol if the connected port is 443. */
-    if( ( ( IotNetworkServerInfo_t * ) ( startInfo.mqttNetworkInfo.u.setup.pNetworkServerInfo ) )->port != 443 )
+    IotLogInfo( "----Device Defender Demo Start----" );
+
+    /* Check parameter(s). */
+    if( ( pIdentifier == NULL ) || ( pIdentifier[ 0 ] == '\0' ) )
     {
-        ( ( IotNetworkCredentials_t * ) ( startInfo.mqttNetworkInfo.u.setup.pNetworkCredentialInfo ) )->pAlpnProtos = NULL;
+        IotLogError( "The length of the Thing Name (identifier) must be nonzero." );
+        status = EXIT_FAILURE;
     }
 
-    /* Set network interface. */
-    startInfo.mqttNetworkInfo.pNetworkInterface = IOT_NETWORK_INTERFACE_AFR;
+    if( status == EXIT_SUCCESS )
+    {
+        /* Initialize the MQTT library. */
+        mqttStatus = IotMqtt_Init();
 
-    /* Set MQTT connection information. */
-    startInfo.mqttConnectionInfo = ( IotMqttConnectInfo_t ) IOT_MQTT_CONNECT_INFO_INITIALIZER;
-    startInfo.mqttConnectionInfo.pClientIdentifier = clientcredentialIOT_THING_NAME;
-    startInfo.mqttConnectionInfo.clientIdentifierLength = ( uint16_t ) strlen( clientcredentialIOT_THING_NAME );
+        if( mqttStatus != IOT_MQTT_SUCCESS )
+        {
+            IotLogError( "MQTT Initialization Failed." );
+            status = EXIT_FAILURE;
+        }
+    }
 
-    /* Set Defender callback. */
-    startInfo.callback = callback;
+    if( status == EXIT_SUCCESS )
+    {
+        /* Initialize Metrics. */
+        metricsInitStatus = IotMetrics_Init();
 
-    /* Invoke defender start API. */
-    defenderResult = AwsIotDefender_Start( &startInfo );
+        if( !metricsInitStatus )
+        {
+            IotLogError( "IOT Metrics Initialization Failed." );
+            status = EXIT_FAILURE;
+        }
+    }
 
-    return defenderResult;
-}
+    if( status == EXIT_SUCCESS )
+    {
+        /* Specify all metrics in "tcp connections" group */
+        defenderResult =
+            AwsIotDefender_SetMetrics( AWS_IOT_DEFENDER_METRICS_TCP_CONNECTIONS, AWS_IOT_DEFENDER_METRICS_ALL );
 
-/*-----------------------------------------------------------*/
+        if( defenderResult != AWS_IOT_DEFENDER_SUCCESS )
+        {
+            status = EXIT_FAILURE;
+        }
+    }
 
-#if _DEMO_WITH_SOCKET_CONNECTED_TO_ECHO_SERVER == 1
+    if( status == EXIT_SUCCESS )
+    {
+        /* Set metrics report period to 5 minutes(300 seconds) */
+        defenderResult = AwsIotDefender_SetPeriod( 300 );
```

Comment:
wouldn't it be better if you use a constant here?

Reply:
Yes.